src/pylast/__init__.py

```diff
@@ -906,68 +913,41 @@ def _download_response(self):
             self.network._delay_call()
 
         username = self.params.pop("username", None)
-        username = f"?username={username}" if username is not None else ""
-
-        data = []
-        for name in self.params.keys():
-            data.append("=".join((name, quote_plus(_string(self.params[name])))))
-        data = "&".join(data)
-
-        headers = {
-            "Content-type": "application/x-www-form-urlencoded",
-            "Accept-Charset": "utf-8",
-            "User-Agent": "pylast/" + __version__,
-        }
+        username = "" if username is None else f"?username={username}"
 
         (host_name, host_subdir) = self.network.ws_server
 
         if self.network.is_proxy_enabled():
-            conn = HTTPSConnection(
-                context=SSL_CONTEXT,
-                host=self.network._get_proxy()[0],
-                port=self.network._get_proxy()[1],
+            client = httpx.Client(
+                verify=SSL_CONTEXT,
+                base_url=f"https://{host_name}",
+                headers=HEADERS,
+                proxies=f"http://{self.network._get_proxy()}",
             )
-
-            try:
-                conn.request(
-                    method="POST",
-                    url=f"https://{host_name}{host_subdir}{username}",
-                    body=data,
-                    headers=headers,
-                )
-            except Exception as e:
-                raise NetworkError(self.network, e) from e
-
         else:
-            conn = HTTPSConnection(context=SSL_CONTEXT, host=host_name)
-
-            try:
-                conn.request(
-                    method="POST",
-                    url=f"{host_subdir}{username}",
-                    body=data,
-                    headers=headers,
-                )
-            except Exception as e:
-                raise NetworkError(self.network, e) from e
+            client = httpx.Client(
+                verify=SSL_CONTEXT,
+                base_url=f"https://{host_name}",
+                headers=HEADERS,
```

Comment:
Looking at this again, it would be ideal to be able to customise the headers sent too. At the very least, it would be idea to have the user agent be customisable.

Reply:
Yes, that could be useful. I'll consider adding an API later. At least they will be customisable by setting `pylast.HEADERS` to something, which wasn't possible before and could be enough.